### PR TITLE
Obj_Syokudai (Torch) OK and mostly documented

### PR DESCRIPTION
**Pull request by rylieb**
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---

This actor's update function was a nightmare!

Pretty interesting actor overall. Key takeaways are:

1) Three different types of torches defined by how they interact with switches (either switches cause them to light up, lighting them up activates a switch, or no interaction with switches at all). Each type has a completely unique DList (gold, black, and brown torches respectively) and the latter two types (which can be lit by the player with FA or lit stick) are COLTYPE_WOOD. This is extremely clever play conditioning! It teaches the player how to interact with each type of torch based on visual cues. For instance, brown torches can sometimes be lit by the player (for example, the torch just up the staircase from the compass room in WFT), but doing so is never required. Black torches lead the player to know it's a torch-lighting puzzle. And gold torches lead the player to know they must complete a puzzle to light them.
2) Likely possible to do a minor exploit with overlay SRM by changing sNumLitTorchesInGroup in the file's BSS to be any number larger than the groupSize. This would activate the switch for lighting all the torches in the group without having to light them yourself.

### CODE REVIEW

File: src/overlays/actors/ovl_Obj_Syokudai/z_obj_syokudai.c [hunk 1]
```diff
@@ -22,38 +32,283 @@ const ActorInit Obj_Syokudai_InitVars = {
...
+        if (groupSizeAdj < this->litTimer) {
+            scaleFactor = ((groupSizeAdj - this->litTimer) + 10) / 10.0f;
+        } else if ((this->litTimer > 0) && (this->litTimer < 20)) {
+            scaleFactor = this->litTimer / 20.0f;
+        }
+        scaleFactor *= 0.0027f;
+        func_8012C2DC(globalCtx->state.gfxCtx);
+        gSPSegment(POLY_XLU_DISP++, 0x08,
+                   Gfx_TwoTexScroll(globalCtx->state.gfxCtx, 0, 0, 0, 0x20, 0x40, 1, 0,
+                                    (this->flameTexScroll * -20) & 0x1FF, 0x20, 0x80));
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Syokudai/z_obj_syokudai.c:304]:
```suggestion
                                    (this->flameTexScroll * -20) % 512, 0x20, 0x80));
```

**Review by engineer124 — COMMENTED** [review]
Nice work with this actor! This file is lit! Just some small suggestions

File: include/functions.h [hunk 2]
```diff
@@ -1109,7 +1109,7 @@ u32 func_800C9E40(CollisionContext* colCtx, CollisionPoly* polygon, s32 index);
 u32 func_800C9E88(CollisionContext* colCtx, CollisionPoly* polygon, s32 index);
 // void func_800C9EBC(GlobalContext* globalCtx, CollisionContext* colCtx, UNK_TYPE1 param_3, UNK_TYPE1 param_4, UNK_TYPE4 param_5, UNK_TYPE4 param_6, UNK_TYPE4 param_7);
 s32 func_800CA1AC(GlobalContext* globalCtx, CollisionContext* colCtx, f32 x, f32 z, f32* ySurface, WaterBox** outWaterBox);
-void func_800CA1E8(GlobalContext* globalCtx, CollisionContext* colCtx, f32 arg2, f32 arg3, f32* arg4, UNK_PTR arg5);
+UNK_TYPE func_800CA1E8(GlobalContext* globalCtx, CollisionContext* colCtx, f32 arg2, f32 arg3, f32* arg4, UNK_PTR arg5);
```
[the remark this answers is not in the document]
Comment (rylieb) [reply include/functions.h:1112]: Yes, I knew this already, but WaterBox is undefined in master, so WaterBox** will surely error out.

File: src/overlays/actors/ovl_Obj_Syokudai/z_obj_syokudai.c [hunk 3]
```diff
@@ -22,38 +32,283 @@ const ActorInit Obj_Syokudai_InitVars = {
...
+    s32 groupSize = OBJ_SYOKUDAI_GET_GROUP_SIZE(thisx);
+    s32 switchFlag = OBJ_SYOKUDAI_GET_SWITCH_FLAG(thisx);
+    s32 type = OBJ_SYOKUDAI_GET_TYPE(thisx);
+    s32 pad0;
+    UNK_PTR waterBox;
+    f32 waterSurface;
+    s32 lightRadius = -1;
+    u8 lightIntensity = 0;
+    Player* player;
+    s32 pad1;
```
[the remark this answers is not in the document]
Comment (rylieb) [reply src/overlays/actors/ovl_Obj_Syokudai/z_obj_syokudai.c:141]: Not that I have been able to figure out. That was what I was trying for the longest time.

File: include/functions.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (engineer124) [reply include/functions.h:1112]: It's in master:
https://github.com/zeldaret/mm/blob/da1265f4bcaec99bc884cb9ee0a5f61535780014/include/z64bgcheck.h#L60-L72

File: src/overlays/actors/ovl_Obj_Syokudai/z_obj_syokudai.c [hunk 4]
```diff
@@ -22,38 +32,283 @@ const ActorInit Obj_Syokudai_InitVars = {
...
+                            this->pendingAction = OBJ_SYOKUDAI_PENDING_ACTION_CUTSCENE_NO_SWITCH;
+                        }
+                    } else if (this->litTimer > 0) {
+                        this->litTimer = -1;
+                    }
+                } else if (this->litTimer < 0) {
+                    this->litTimer = 20;
+                }
+            }
+            if (this->flameCollider.base.acFlags & AC_HARD) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Syokudai/z_obj_syokudai.c:192]:
```suggestion
            if (this->flameCollider.base.acFlags & AC_HIT) {
```
I made a dumb dumb in my review

File: src/overlays/actors/ovl_Obj_Syokudai/z_obj_syokudai.c (hunk 1 above)
Comment (rylieb) [reply src/overlays/actors/ovl_Obj_Syokudai/z_obj_syokudai.c:304]: Does not match.

File: src/overlays/actors/ovl_Obj_Syokudai/z_obj_syokudai.h [hunk 5]
```diff
@@ -7,12 +7,32 @@ struct ObjSyokudai;
...
+    /* 0 */ OBJ_SYOKUDAI_TYPE_SWITCH_CAUSES_FLAME,
+    /* 1 */ OBJ_SYOKUDAI_TYPE_FLAME_CAUSES_SWITCH,
+    /* 2 */ OBJ_SYOKUDAI_TYPE_NO_SWITCH
+} ObjSyokudaiTypes;
+
+typedef enum {
+    /* -1 */ OBJ_SYOKUDAI_PENDING_ACTION_CUTSCENE_NO_SWITCH = -1,
+    /*  0 */ OBJ_SYOKUDAI_PENDING_ACTION_NONE,
+    /*  1 */ OBJ_SYOKUDAI_PENDING_ACTION_CUTSCENE_AND_SWITCH
+} ObjSyokudaiPendingActions;
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Obj_Syokudai/z_obj_syokudai.h:31]: I wonder if we could chop of a few words of this :thinking:  Maybe remove the "PENDING"?
```suggestion
typedef enum {
    /* -1 */ OBJ_SYOKUDAI_ACTION_CUTSCENE_NO_SWITCH = -1,
    /*  0 */ OBJ_SYOKUDAI_ACTION_NONE,
    /*  1 */ OBJ_SYOKUDAI_ACTION_CUTSCENE_AND_SWITCH
} ObjSyokudaiPendingActions;
```

Comment (rylieb) [reply src/overlays/actors/ovl_Obj_Syokudai/z_obj_syokudai.h:31]: I feel like pending describes it better, personally. The pending action gets set at the bottom of the update func, and then the top of the func checks for a pending action, executes it, and then sets the pending action to NONE. But I do agree the names are a little long.

Comment (rylieb) [reply src/overlays/actors/ovl_Obj_Syokudai/z_obj_syokudai.h:31]: Alternatively, i could get rid of "_NO_SWITCH" at the end of the first one and just have it be implicit. I thought about doing it, but I figured making it clear that this was the nonswitching case was more important than brevity.

**Review by AngheloAlf — APPROVED** [review]

**Review by engineer124 — APPROVED** [review]

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**